---
title: "GO_KEGG_pathway"
author: "Kelly Ren"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::html_document2: default
  ---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r}
library(ggplot2)
library(easyGgplot2)
library(magrittr)
library(readr)
library(dplyr)
library(stringr)
library(tidyverse)
library(readxl)
library(gtools)
library(tibble)
library(reshape2)
library(GenomicFeatures)

'%!in%' <- function(x,y)!('%in%'(x,y))
```

```{r}
 setwd("~/Documents/Kelly_2020/Research/Six_M_Work_2020/Kelly_Buffalo_Paper/Buffalo_Frontier_genetics_v2")
```

# Test different model

Use R script to plot the distribution of 2*(global_lnLâˆ’multi_lnL), so prove the globallambda
Rscript lhtest.R lh2to1.out.edit -52526.384355 -51293.524994

# error correction

```{r}
caferror_default_output <- read.table("caferror_default_output.txt",header = T)
min_Score <- min(caferror_default_output$Score)
subset(caferror_default_output, Score %in% min_Score) 
# Best error model 0.02597656
```

# CAFE results
## human
```{r}
Gene_GL_PTHR <- read.table("Gene_GL_PTHR.txt")%>%
  set_colnames("PANTHER_ID")

match_table_species <- read.csv("match_table_species.csv")

Gene_GL_PTHR%>%
  left_join(match_table_species)%>%
  extract2("UniProtKB")%>%
  unique()%>%
  write.csv("results_Gene_GL_PTHR.csv")

# match UniportKB to gene on website
Gene_GL_PTHR%>%
  left_join(match_table_species)%>%
  write.csv("results_Gene_GL_PTHR.csv")
```


```{r}
UniProtKB_toEntrez_geneGL <- read.table("UniProtKB_toEntrez_geneGL.txt",header = T)
UniProtKB_toGenename_geneGL <- read.table("UniProtKB_toGenename_geneGL.txt",header = T)

results_Gene_GL_PTHR <- Gene_GL_PTHR%>%
  left_join(match_table_species, by ="PANTHER_ID")%>%
  left_join(UniProtKB_toEntrez_geneGL, by = "UniProtKB")%>%
  left_join(UniProtKB_toGenename_geneGL, by = "UniProtKB")

write.csv(results_Gene_GL_PTHR,"results_Gene_GL_PTHR.csv")
```

```{r}
#ah <- AnnotationHub()
#saveRDS(ah,"All_annotation.rds")
ah <- read_rds("All_annotation.rds") 
#subset(ah, rdataclass == "EnsDb" & species == "Homo sapiens")
ensDb <- ah[["AH83216"]]
ensDb
```

```{r}
genesGR <- GenomicFeatures::genes(ensDb)
genesGR
```

```{r }
cols2Keep <- c("gene_id","gene_name", "gene_biotype", "description", "entrezid")
mcols(genesGR) <- mcols(genesGR)[, cols2Keep]

Genes <- genesGR%>%
  as.data.frame()
Genes$entrezid <- Genes$entrezid%>%as.character()
```


```{r}
results_Gene_GL_PTHR$ToEntrez <- results_Gene_GL_PTHR$ToEntrez%>%as.character()
results_Gene_GL_PTHR <- results_Gene_GL_PTHR%>%
  left_join(Genes, by = c("ToEntrez" = "entrezid"))%>%
  subset(gene_biotype %in% "protein_coding")
```

## Search immune genes for Human in immune database

```{r}
results_Gene_GL_PTHR <- results_Gene_GL_PTHR%>%subset(species %in% "Hsap")
```


```{r}
InnateDB_immune <- read_csv("InnateDB_genes_all.csv")%>%
  as.data.frame()
InnateDB_immune$entrez <- as.character(InnateDB_immune$entrez)

InnateDB_immune$species%>%
  table() #species Homo sapiens and mouse
InnateDB_immune_human <- subset(InnateDB_immune,species %in% "Homo sapiens")
selected_InnateDB_immune_human <- InnateDB_immune_human[,c("id","species","taxonId","ensembl","entrez","name","fullname","synonym","signature","chromStart", "chromEnd", "chromStrand", "chromName", "goTerms","function","humanOrthologs","mouseOrthologs","bovineOrthologs")]

selected_InnateDB_immune_human_MHCGRanges <- selected_InnateDB_immune_human%>%set_colnames(c("id","species","taxonId","ensembl","entrez","name","fullname","synonym","signature","start", "end", "strand", "chr", "goTerms","function","humanOrthologs","mouseOrthologs","bovineOrthologs"))%>%
  makeGRangesFromDataFrame()

mcols(selected_InnateDB_immune_human_MHCGRanges) <- selected_InnateDB_immune_human

results_Gene_GL_PTHR_GRanges <- makeGRangesFromDataFrame(results_Gene_GL_PTHR)
mcols(results_Gene_GL_PTHR_GRanges) <- results_Gene_GL_PTHR

# Overlaps to find the immune genes
results_Gene_GL_PTHR_immune <- results_Gene_GL_PTHR_GRanges[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,selected_InnateDB_immune_human_MHCGRanges)),]

write_csv(as.data.frame(mcols(results_Gene_GL_PTHR_immune)),"results_Gene_GL_PTHR_immune.csv")

# Found MHC II
results_Gene_GL_PTHR_immune[mcols(results_Gene_GL_PTHR_immune)$description%>%grep("histocompatibility",.),]

results_Gene_GL_PTHR_immune[mcols(results_Gene_GL_PTHR_immune)$description%>%grep("histocompatibility",.),]%>%
  mcols()%>%
  extract2("PANTHER_ID")%>%
  unique()

results_Gene_GL_PTHR_immune[mcols(results_Gene_GL_PTHR_immune)$description%>%grep("histocompatibility",.),]%>%
  mcols()%>%
  extract2("gene_name")%>%
  unique()

results_Gene_GL_PTHR_immune[mcols(results_Gene_GL_PTHR_immune)$description%>%grep("histocompatibility",.),]%>%
  mcols()%>%
  extract2("orthogroup_ID")%>%
  unique() 
```

## Search Callum immune genes

```{r}
MHC_species_gene_coordinate <- read_csv("MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])
# Search for human only
mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])[subjectHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]


mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]%>%
  extract2("PANTHER_ID")%>%
  unique()

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])[subjectHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]%>%
  as.data.frame()%>%
  extract2("gene_ID")

mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]%>%
  extract2("UniProtKB")%>%
  unique()%>%
  as.data.frame()
```

## cattle

```{r}
#ah <- AnnotationHub()
#saveRDS(ah,"All_annotation.rds")
ah <- read_rds("All_annotation.rds") 
#subset(ah, rdataclass == "EnsDb" & species == "Homo sapiens")
ensDb <- ah[["AH83145"]]
ensDb
```

```{r}
genesGR <- genes(ensDb)
genesGR
```

```{r }
cols2Keep <- c("gene_id","gene_name", "gene_biotype", "description", "entrezid")
mcols(genesGR) <- mcols(genesGR)[, cols2Keep]

Genes <- genesGR%>%
  as.data.frame()
Genes$entrezid <- Genes$entrezid%>%
  as.character()
```


## Search Callum immune genes

```{r}
results_Gene_GL_PTHR <- read_rds("results_Gene_GL_PTHR.rds")
results_Gene_GL_PTHR <- results_Gene_GL_PTHR%>%subset(species %in% "Hbta")
results_Gene_GL_PTHR_GRanges <- makeGRangesFromDataFrame(results_Gene_GL_PTHR)
mcols(results_Gene_GL_PTHR_GRanges) <- results_Gene_GL_PTHR
```

```{r}
MHC_species_gene_coordinate <- read_csv("MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "cattle",])
# Search for cattle only
mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "cattle",])),]

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "cattle",])[subjectHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "cattle",])),]


mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "cattle",])),]%>%
  extract2("PANTHER_ID")%>%
  unique()
```

## goat

```{r}
#ah <- AnnotationHub()
#saveRDS(ah,"All_annotation.rds")
ah <- read_rds("All_annotation.rds") 
#subset(ah, rdataclass == "EnsDb" & species == "Capra hircus")
ensDb <- ah[["AH83158"]]
ensDb
```

```{r}
genesGR <- genes(ensDb)
genesGR
```

```{r }
cols2Keep <- c("gene_id","gene_name", "gene_biotype", "description", "entrezid")
mcols(genesGR) <- mcols(genesGR)[, cols2Keep]

Genes <- genesGR%>%
  as.data.frame()
Genes$entrezid <- Genes$entrezid%>%
  as.character()
```


## Search Callum immune genes

```{r}
results_Gene_GL_PTHR <- read_rds("results_Gene_GL_PTHR.rds")
results_Gene_GL_PTHR <- results_Gene_GL_PTHR%>%subset(species %in% "Chir")
results_Gene_GL_PTHR_GRanges <- makeGRangesFromDataFrame(results_Gene_GL_PTHR)
mcols(results_Gene_GL_PTHR_GRanges) <- results_Gene_GL_PTHR
```

```{r}
MHC_species_gene_coordinate <- read_csv("MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "goat",])
# Search for goat only
mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "goat",])),]

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "goat",])[subjectHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "goat",])),]


mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "goat",])),]%>%
  extract2("PANTHER_ID")%>%
  unique()
```

## pig

```{r}
#ah <- AnnotationHub()
#saveRDS(ah,"All_annotation.rds")
ah <- read_rds("All_annotation.rds") 
#subset(ah, rdataclass == "EnsDb" & species == "Sus scrofa")
ensDb <- ah[["AH83340"]]
ensDb
```

```{r}
genesGR <- genes(ensDb)
genesGR
```

```{r }
cols2Keep <- c("gene_id","gene_name", "gene_biotype", "description", "entrezid")
mcols(genesGR) <- mcols(genesGR)[, cols2Keep]

Genes <- genesGR%>%
  as.data.frame()
Genes$entrezid <- Genes$entrezid%>%
  as.character()
```


## Search Callum immune genes

```{r}
results_Gene_GL_PTHR <- read_rds("results_Gene_GL_PTHR.rds")
results_Gene_GL_PTHR <- results_Gene_GL_PTHR%>%subset(species %in% "Sscr")
results_Gene_GL_PTHR_GRanges <- makeGRangesFromDataFrame(results_Gene_GL_PTHR)
mcols(results_Gene_GL_PTHR_GRanges) <- results_Gene_GL_PTHR
```

```{r}
MHC_species_gene_coordinate <- read_csv("MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "pig",])
# Search for pig only
mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "pig",])),]

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "pig",])[subjectHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "pig",])),]


mcols(results_Gene_GL_PTHR_GRanges)[queryHits(findOverlaps(results_Gene_GL_PTHR_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "pig",])),]%>%
  extract2("PANTHER_ID")%>%
  unique()
```

