---
title: "PAML_Find_immune_Gene"
author: "Kelly Ren"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  bookdown::html_document2: default
bibliography: FinalthesisRef.bib
link-citations: yes
fig_caption: yes
toc: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

# Packages

```{r}
library(magrittr)
library(readr)
library(dplyr)
library(stringr)
library(tidyverse)
library(readxl)
library(gtools)
library(tibble)
library(AnnotationHub)
library(biomaRt)
```

# Analysis

```{r}
genes_in_orthogroup_all_species <- read_csv("Output/CAFE/genes_in_orthogroup_all_species.csv")
```

## Positive select results
### ref Human database

```{r}
human_results_Positive_select <- read_csv("Output/PAML/branch_results_Positive_select.csv")%>%
  as.data.frame()

# human only (The matching table is majoritily base on human)
human_results_Positive_select$species%>%table()

# number of genes 602
human_results_Positive_select$entrezgene_id%>%unique%>%length()

write_csv(as.data.frame(human_results_Positive_select),"Output/PAML/humanref_results_Positive_select.csv")
```


```{r}
#ah <- AnnotationHub()
#saveRDS(ah,"All_annotation.rds")
ah <- read_rds("/Users/kellydren/Documents/Kelly_annotation/All_annotation.rds") 
#subset(ah, rdataclass == "EnsDb" & species == "Homo sapiens")
ensDb <- ah[["AH83216"]]
ensDb
```

```{r}
genesGR <- genes(ensDb)
genesGR
```

```{r }
cols2Keep <- c("gene_id","gene_name", "gene_biotype", "description", "entrezid")
mcols(genesGR) <- mcols(genesGR)[, cols2Keep]

Genes <- genesGR%>%
  as.data.frame()
Genes$entrezid <- Genes$entrezid%>%as.character()
```


```{r}
human_results_Positive_select$entrezgene_id <- human_results_Positive_select$entrezgene_id%>%as.character()
human_results_Positive_select <- human_results_Positive_select%>%
  left_join(Genes, by = c("entrezgene_id" = "entrezid"))%>%
  subset(gene_biotype %in% "protein_coding")
colnames(human_results_Positive_select) <- colnames(human_results_Positive_select)%>%gsub("entrezgene_id","entrez",.)
```

#### Search immune genes for Human in immune database

```{r}
# This table downloaded from InnateDB immune database for human, mouse only
InnateDB_immune <- read_csv("Input/Immune_gene_ref/InnateDB_genes_all.csv")%>%
  as.data.frame()
InnateDB_immune$entrez <- as.character(InnateDB_immune$entrez)

InnateDB_immune$species%>%
  table() #species Homo sapiens and mouse

# selected for human only
InnateDB_immune_human <- subset(InnateDB_immune,species %in% "Homo sapiens")
selected_InnateDB_immune_human <- InnateDB_immune_human[,c("id","species","taxonId","ensembl","entrez","name","fullname","synonym","signature","chromStart", "chromEnd", "chromStrand", "chromName", "goTerms","function","humanOrthologs","mouseOrthologs","bovineOrthologs")]

selected_InnateDB_immune_human_MHCGRanges <- selected_InnateDB_immune_human%>%set_colnames(c("id","species","taxonId","ensembl","entrez","name","fullname","synonym","signature","start", "end", "strand", "chr", "goTerms","function","humanOrthologs","mouseOrthologs","bovineOrthologs"))%>%
  makeGRangesFromDataFrame()

mcols(selected_InnateDB_immune_human_MHCGRanges) <- selected_InnateDB_immune_human

human_results_Positive_select_GRanges <- makeGRangesFromDataFrame(human_results_Positive_select)
mcols(human_results_Positive_select_GRanges) <- human_results_Positive_select

# Overlaps to find the immune genes
human_results_Positive_select_immune <- human_results_Positive_select_GRanges[queryHits(findOverlaps(human_results_Positive_select_GRanges,selected_InnateDB_immune_human_MHCGRanges)),]

human_results_Positive_select_immune$ID%>%unique()%>%length()

# output the Positive_select results for immune gene only
write_csv(as.data.frame(mcols(human_results_Positive_select_immune)),"Output/PAML/humanref_results_Positive_select_immune.csv")
```


```{r}
# Check if result contain MHC I or II
human_results_Positive_select_immune[mcols(human_results_Positive_select_immune)$description%>%grep("histocompatibility",.),]

human_results_Positive_select_immune[mcols(human_results_Positive_select_immune)$description%>%grep("histocompatibility",.),]%>%mcols()%>%
  extract2("ID")%>%
  unique() 

human_results_Positive_select_immune[mcols(human_results_Positive_select_immune)$description%>%grep("histocompatibility",.),]%>%mcols()%>%
  extract2("ID")%>%
  unique() 
```

#### Search through our manully identified MHC immune genes

```{r}
MHC_species_gene_coordinate <- read_csv("Input/Immune_gene_ref/Manul_MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

# Search for human only
mcols(human_results_Positive_select_GRanges)[queryHits(findOverlaps(human_results_Positive_select_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]

mcols(MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])[subjectHits(findOverlaps(human_results_Positive_select_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "human",])),]
```

### Bbub
```{r}
Bbub_results_Positive_select <- read_csv("Output/PAML/branch_results_Positive_select.csv")%>%
  as.data.frame()

Bbub_results_Positive_select <- genes_in_orthogroup_all_species%>%
  subset(spc %in% "Bbub")%>%
  subset(orthogroup_ID %in% Bbub_results_Positive_select$orthogroup_ID)
```

```{r}
# The gff file of river buffalo can be download from 
river_buffalogff <- read_rds("Input/Species_annotation/river_buffalogff.rds")

river_buffalogff <- subset(river_buffalogff, type %in% "gene")
river_buffalogff$entrezID <- river_buffalogff$attributes%>%gsub("^.*GeneID:","",.)%>%gsub(";Name=.*","",.)

#seqid names not in "1","2", ...
#read in ncbi assembly full seq report that has been processed to fix this
river_buffalo_rename <- read_delim("Input/Babesia_tblastn_output_20200901/river_buffalo.rename.csv",",",col_names = FALSE)

river_buffalogff <- river_buffalogff[grep("protein_coding", river_buffalogff[,9]),]%>%
  unique()

# Change the chr names
for (i in 1:nrow(river_buffalo_rename)){
river_buffalogff$seqid <- gsub(river_buffalo_rename$X2[i],river_buffalo_rename$X1[i],river_buffalogff$seqid)}

river_buffaloGRanges <- makeGRangesFromDataFrame(river_buffalogff)%>%unique()
river_buffalogff$seqid <- river_buffalogff$seqid%>%
  gsub(" X", "X",.)  

mcols(river_buffaloGRanges) <- river_buffalogff

# subset a GRanges for positive selected Bbub
Bbub_results_Positive_select_GRanges<- subset(river_buffaloGRanges, entrezID %in% Bbub_results_Positive_select$geneID)
```

#### Search through our manully identified MHC immune genes

```{r}
MHC_species_gene_coordinate <- read_csv("Input/Immune_gene_ref/Manul_MHC_species_gene_coordinate_table.csv")%>%
  as.data.frame()

MHC_species_gene_coordinate_MHCGRanges <- makeGRangesFromDataFrame(MHC_species_gene_coordinate)
mcols(MHC_species_gene_coordinate_MHCGRanges) <- MHC_species_gene_coordinate

# Search for Bbub only
hit_MHC_Bbub <- findOverlaps(Bbub_results_Positive_select_GRanges,MHC_species_gene_coordinate_MHCGRanges[mcols(MHC_species_gene_coordinate_MHCGRanges)$species %in% "Bbub",])

length(hit_MHC_Bbub)

mcols(Bbub_results_Positive_select_GRanges)[queryHits(hit_MHC_Bbub),]%>%as.data.frame()

# MHC that in the positive selected othrogroups
mcols(MHC_species_gene_coordinate_MHCGRanges)[subjectHits(hit_MHC_Bbub),]%>%as.data.frame()%>%extract2("gene_symbol")%>%unique()
```

### ref Cattle database

```{r}
# Selcet the genes in cattle orthogroups only
Hbta_results_Positive_select <- read_csv("Output/PAML/branch_results_Positive_select.csv")%>%
  as.data.frame()

Hbta_results_Positive_select <- genes_in_orthogroup_all_species%>%
  subset(spc %in% "Hbta")%>%
  subset(orthogroup_ID %in%
Hbta_results_Positive_select$orthogroup_ID)

Hbta_results_Positive_select$geneID%>%unique()%>%length()

# Selcet the genes in buffalo orthogroups only
Bbub_results_Positive_select <- read_csv("Output/PAML/branch_results_Positive_select.csv")%>%
  as.data.frame()

Bbub_results_Positive_select <- genes_in_orthogroup_all_species%>%
  subset(spc %in% "Bbub")%>%
  subset(orthogroup_ID %in%
Bbub_results_Positive_select$orthogroup_ID)

# Select for buffalo only 
Hbta_results_Positive_select <- subset(Hbta_results_Positive_select,orthogroup_ID %in% Bbub_results_Positive_select$orthogroup_ID)

# 588 genes left
Hbta_results_Positive_select$geneID%>%unique()%>%length()

write_csv(as.data.frame(Hbta_results_Positive_select),"Output/PAML/Hbtaref_results_Positive_select.csv")
  
```

# Appendix

```{r}
sessionInfo()
```